### ISSUE TITLE
Stepup Problem in Firefox

### ISSUE THREAD

**Issue by gergan** [zmartzone/mod_auth_openidc#404]
Hi, 
we are using 
- mod_auth_openidc master (HEAD - 2.3.9rcX - c30d3e8fdc9d696727b3af1004a4fe6d29c02af9)
- Apache version mod_ssl/2.4.27 compiled against Server: Apache/2.4.27, Library: OpenSSL/1.0.1e
- Red Hat 7
in a docker container.
We have a problem with the Stepup in Firefox, at least on a situation, in which the user is already logged in, but does not have the right permissions, the pages flicker between the redirect to OP (keycloak in the this case).

The Response has the redirect-page and the 401 response page in one:
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>Stepup Authentication</title>
    <meta http-equiv="refresh" content="0; url=https://coreos-grow:4664/auth/realms/customer/protocol/openid-connect/auth?response_type=code&scope=openid%20profile%20email%20private-customer-portal&client_id=private-customer-portal&state=JshP54OkTMtk2GRFsJjfniiOXcE&redirect_uri=https%3A%2F%2Fcoreos-grow%3A8443%2Fredirect_uri&nonce=FzDEnELEJ9IJz7zWlLPISDv_4XBrTdvTY7I1J-PHwAQ&claims=%7B%22id_token%22%3A%20%7B%22acr%22%3A%7B%22essential%22%3A%20true%2C%20%22values%22%3A%5B%222%22%5D%7D%7D%7D">
  </head>
  <body>
<p></p>
  </body>
</html>
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>401 Unauthorized</title>
</head><body>
<h1>Unauthorized</h1>
<p>This server could not verify that you
are authorized to access the document
requested.  Either you supplied the wrong
credentials (e.g., bad password), or your
browser doesn't understand how to supply
the credentials required.</p>
</body></html>
```

![login_which_needs_stepup](https://user-images.githubusercontent.com/444543/48269919-069cac80-e439-11e8-95b4-2a979eccb66e.png)

die Logs from apche:

```
[Fri Nov 09 12:25:48.007454 2018] [auth_openidc:debug] [pid 65] src/util.c(2285): [client 172.18.0.3:57946] oidc_util_hdr_table_set: Location: https://coreos-grow:4664/auth/realms/custo
mer/protocol/openid-connect/auth?response_type=code&scope=openid%20profile%20email%20private-customer-portal&client_id=private-customer-portal&state=JshP54OkTMtk2GRFsJjfniiOXcE&redirect
_uri=https%3A%2F%2Fcoreos-grow%3A8443%2Fredirect_uri&nonce=FzDEnELEJ9IJz7zWlLPISDv_4XBrTdvTY7I1J-PHwAQ&claims=%7B%22id_token%22%3A%20%7B%22acr%22%3A%7B%22essential%22%3A%20true%2C%20%22
values%22%3A%5B%222%22%5D%7D%7D%7D
[Fri Nov 09 12:25:48.007459 2018] [auth_openidc:debug] [pid 65] src/util.c(2307): [client 172.18.0.3:57946] oidc_util_hdr_err_out_add: Cache-Control: no-cache, no-store, max-age=0
[Fri Nov 09 12:25:48.007461 2018] [auth_openidc:debug] [pid 65] src/proto.c(689): [client 172.18.0.3:57946] oidc_proto_authorization_request: return: 302
[Fri Nov 09 12:25:48.007463 2018] [auth_openidc:debug] [pid 65] src/mod_auth_openidc.c(3731): [client 172.18.0.3:57946] oidc_handle_unauthorized_user24: send HTML refresh with authoriza
tion redirect: https://coreos-grow:4664/auth/realms/customer/protocol/openid-connect/auth?response_type=code&scope=openid%20profile%20email%20private-customer-portal&client_id=private-c
ustomer-portal&state=JshP54OkTMtk2GRFsJjfniiOXcE&redirect_uri=https%3A%2F%2Fcoreos-grow%3A8443%2Fredirect_uri&nonce=FzDEnELEJ9IJz7zWlLPISDv_4XBrTdvTY7I1J-PHwAQ&claims=%7B%22id_token%22%
3A%20%7B%22acr%22%3A%7B%22essential%22%3A%20true%2C%20%22values%22%3A%5B%222%22%5D%7D%7D%7D
[Fri Nov 09 12:25:48.007491 2018] [http:trace3] [pid 65] http_filters.c(1128): [client 172.18.0.3:57946] Response sent with status 200, headers:
[Fri Nov 09 12:25:48.007493 2018] [http:trace5] [pid 65] http_filters.c(1135): [client 172.18.0.3:57946]   Date: Fri, 09 Nov 2018 12:25:48 GMT
[Fri Nov 09 12:25:48.007495 2018] [http:trace5] [pid 65] http_filters.c(1138): [client 172.18.0.3:57946]   Server: Apache/2.4.27 (Red Hat) OpenSSL/1.0.1e-fips
[Fri Nov 09 12:25:48.007498 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Set-Cookie: mod_auth_openidc_session=7a3e1ecb-9e47-4ac5-8856-3e95c36df8f2; Path
=/; Secure; HttpOnly
[Fri Nov 09 12:25:48.007500 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Set-Cookie: mod_auth_openidc_state_Vdw7JmkAzKjF52hNWtI3Nlu2T90=; Path=/; Expire
s=Thu, 01 Jan 1970 00:00:00 GMT; Secure; HttpOnly
[Fri Nov 09 12:25:48.007502 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Set-Cookie: mod_auth_openidc_state_JshP54OkTMtk2GRFsJjfniiOXcE=eyJhbGciOiAiZGly
IiwgImVuYyI6ICJBMjU2R0NNIn0..nuwJLxDlEIhaKTsj.mtI3ORMeLQvgnNm8JQUG_fyYZlkQOVddnDLd8ZZrBbhFcfgGxKSdkBHrHXZxiDHzqT_7YyMIna3OIDQ3wr8v-78XB6xSXeG3PS9kcF_j7DdwxQhNkaVlc03liVToDWHidMf0bIWhRyh
0KSnz-cKSPdOcQb3gPht8-5lBMeYDzx7WuVQXJFUMI1Hj9ps3O86uJqp8dI8sP_DRLQpuze1FFGhZPdbdYRd_Jpke5RfgakU937j_TQ_YspJmEPRL5aYAcescSwuR4EOWh-NIGpN8bk8Ys7wCuU-hvZtqFm3KkjT2F5tlwDX5w1073ZKHM96O1p-x
Gjq4PlqAyWEvdaO7ClMVZwnef74bbNKOkrMlMExSklgt5ZkOGzeGKK0fkHjJHskStux13ZQ-d0RpxhbOlxqB2ePLkBI.x6AT4e68bSoGDgEC4anTMQ; Path=/; Secure; HttpOnly
[Fri Nov 09 12:25:48.007504 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Cache-Control: no-cache, no-store, max-age=0
[Fri Nov 09 12:25:48.007505 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Location: https://coreos-grow:4664/auth/realms/customer/protocol/openid-connect
/auth?response_type=code&scope=openid%20profile%20email%20private-customer-portal&client_id=private-customer-portal&state=JshP54OkTMtk2GRFsJjfniiOXcE&redirect_uri=https%3A%2F%2Fcoreos-g
row%3A8443%2Fredirect_uri&nonce=FzDEnELEJ9IJz7zWlLPISDv_4XBrTdvTY7I1J-PHwAQ&claims=%7B%22id_token%22%3A%20%7B%22acr%22%3A%7B%22essential%22%3A%20true%2C%20%22values%22%3A%5B%222%22%5D%7
D%7D%7D
[Fri Nov 09 12:25:48.007507 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Content-Length: 761
[Fri Nov 09 12:25:48.007508 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Connection: close
[Fri Nov 09 12:25:48.007510 2018] [http:trace4] [pid 65] http_filters.c(957): [client 172.18.0.3:57946]   Content-Type: text/html; charset=UTF-8
[Fri Nov 09 12:25:48.007647 2018] [authz_core:debug] [pid 65] mod_authz_core.c(809): [client 172.18.0.3:57946] AH01626: authorization result of Require claims_expr '.authorities | index
("pcp-customer-stepup") != null': denied
[Fri Nov 09 12:25:48.007653 2018] [authz_core:debug] [pid 65] mod_authz_core.c(809): [client 172.18.0.3:57946] AH01626: authorization result of <RequireAny>: denied
[Fri Nov 09 12:25:48.007655 2018] [authz_core:error] [pid 65] [client 172.18.0.3:57946] AH01631: user t@t.de: authorization failure for "/dashboard":
[Fri Nov 09 12:25:48.007660 2018] [core:trace3] [pid 65] request.c(119): [client 172.18.0.3:57946] auth phase 'check authorization' gave status 401: /dashboard
[Fri Nov 09 12:25:48.007680 2018] [core:trace6] [pid 65] core_filters.c(525): [client 172.18.0.3:57946] core_output_filter: flushing because of FLUSH bucket
172.18.0.3 - t@t.de [09/Nov/2018:12:25:48 +0000] "GET /dashboard HTTP/1.1" 401 1142 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:58.0) Gecko/20100101 Firefox/58.0"
[Fri Nov 09 12:25:48.007835 2018] [core:trace6] [pid 65] core_filters.c(525): [client 172.18.0.3:57946] core_output_filter: flushing because of FLUSH bucket
```

I think the problem is in oidc_handle_unauthorized_user24:

```
	oidc_authenticate_user(r, c, NULL, oidc_get_current_url(r), NULL,
			NULL, NULL, oidc_dir_cfg_path_auth_request_params(r),
			oidc_dir_cfg_path_scope(r));

	const char *location = oidc_util_hdr_out_location_get(r);
	if (location != NULL) {
		oidc_debug(r, "send HTML refresh with authorization redirect: %s",
				location);

		char *html_head = apr_psprintf(r->pool,
				"<meta http-equiv=\"refresh\" content=\"0; url=%s\">",
				location);
		oidc_util_html_send(r, "Stepup Authentication", html_head, NULL, NULL,
				HTTP_UNAUTHORIZED);
	}

	return AUTHZ_DENIED;
```

But it could be that our configuration is in some way strange:
```
    <Location ~ "/*">
        AuthType openid-connect
        Require claims_expr '.authorities | index("pcp-customer") != null'
        RequestHeader set Authorization "Bearer %{OIDC_access_token}e" env=OIDC_access_token
        OIDCPathScope "private-customer-portal"    
        ProxyPass "${UPSTREAM_URL}"
        ProxyPassReverse "${UPSTREAM_URL}"
        RequestHeader set Authorization "Bearer %{OIDC_access_token}e" env=OIDC_access_token
        AddOutputFilterByType SUBSTITUTE text/html
        Substitute "s|</body>|<iframe title='empty' style='visibility: hidden;' width='0' height='0' tabindex='-1' id='openidc-op' src='${KEYCLOAK_PROXY_REDIRECTION_URL}?session=iframe_op'></iframe><iframe title='empty' style='visibility: hidden;' width='0' height='0' tabindex='-1' id='openidc-rp' src='${KEYCLOAK_PROXY_REDIRECTION_URL}?session=iframe_rp\&amp;poll=5000'></iframe></body>|i"
    </Location>

    <LocationMatch "^(?!/test)/[^/]+">
        Require claims_expr '.authorities | index("pcp-customer-stepup") != null'
        OIDCPathAuthRequestParams claims=%7B%22id_token%22%3A%20%7B%22acr%22%3A%7B%22essential%22%3A%20true%2C%20%22values%22%3A%5B%222%22%5D%7D%7D%7D
        OIDCUnAutzAction auth
        OIDCPathScope "private-customer-portal"
    </LocationMatch>
```

Chrome/Opera seem to stop processing the response, right after the first closing </html>-Tag and in this way does not show the 401-Page simply redirecting to keycloak.

**Comment by zandbelt** [comment]
Thanks for the detailed report, that makes it easy to look into it!

Interestingly enough I cannot reproduce this on my Apache 2.4.37  environment: it doesn't include the "401 Unauthorized" HTML part.

Could you try and set `ErrorDocument 401 ""`

**Comment by gergan** [comment]
Ok ErrorDocument 401 " " seems to fix the problem (apparently there should be a space there, otherwise apache complains).
Should I close the issue or is the setting more of a workaround?

**Comment by zandbelt** [comment]
I've added a note here https://github.com/zmartzone/mod_auth_openidc/wiki/Step-up-Authentication thanks!

**Comment by candlerb** [comment]
Just to clarify, does it need to be exactly
```
ErrorDocument 401 " "
```
or would
```
ErrorDocument 401 "You are not authorized"
```
work sufficiently well?

Using the former, if the user is not authorized they just see a blank white page, which isn't very helpful.

**Comment by candlerb** [comment]
I can no longer reproduce this issue with current Firefox (97.0.1, macOS) and mod_auth_openidc 2.4.1 (Ubuntu 20.04)

If I remove the ErrorDocument line, and have a user who authenticates but is not authorized (does not meet a `Require claim groups:XXX`, I just get the expected Apache error page:

-----

## Unauthorized

This server could not verify that you are authorized to access the document requested. Either you supplied the wrong credentials (e.g., bad password), or your browser doesn't understand how to supply the credentials required.

*Apache/2.4.41 (Ubuntu) Server at xxxxxx Port 443*

-----

Page source:

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>401 Unauthorized</title>
</head><body>
<h1>Unauthorized</h1>
<p>This server could not verify that you
are authorized to access the document
requested.  Either you supplied the wrong
credentials (e.g., bad password), or your
browser doesn't understand how to supply
the credentials required.</p>
<hr>
<address>Apache/2.4.41 (Ubuntu) Server at netbox.home.deploy2.net Port 443</address>
</body></html>
```

There's no meta redirect there.

I know I *did* suffer from this problem previously, because I bookmarked this issue and updated my configs with `ErrorDocument 401 " "` to fix it; but I was using Ubuntu 18.04 with an older version of mod_auth_openidc (2.3.3) and Apache (2.4.29)

**Comment by candlerb** [comment]
I think I have the answer now:

```
    <Location />
        AuthType openid-connect
        Require claim ....
        AuthzSendForbiddenOnFailure on
    </Location>
```

This tells Apache to give a 403 Forbidden response, which makes it clear to the end user what the issue is, and should ensure there's no possibility of a redirect loop.
